Link For the dataset: [LINK](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/green/green_tripdata_2019-09.csv.gz)

In [1]:
import pandas as pd
pd.__version__

/tmp/ipykernel_13528/2621542425.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


'2.2.0'

### We are gonna use the `sqlalchemy` module for connecting to postgresql database and inserting the data.

In [7]:
# Importing the module
from sqlalchemy import create_engine

#### Creating connection with postgres db.

In [8]:
# Credentials for connecting to db
user_name = "root"
password = "root"
database_name = "ny_taxi"
host = "localhost"
port = "5432"

# Connection string format
connection_string = f"postgresql://{user_name}:{password}@{host}:{port}/{database_name}"

engine = create_engine(connection_string)

# Testing connection
try:
    with engine.connect() as connection_str:
        print('Successfully connected to the PostgreSQL database')
except Exception as ex:
    print(f'Sorry failed to connect: {ex}')

Successfully connected to the PostgreSQL database


#### Counting the number of rows available inside our dataset.

In [9]:
!wc -l green_tripdata_2019-09.csv.gz

29689 green_tripdata_2019-09.csv.gz


#### Our dataset contains 29689 number of rows so for smooth operation we are gonna divide the dataset into multiple chunks.

In [36]:
df_iter = pd.read_csv("green_tripdata_2019-09.csv.gz", iterator=True, chunksize=30000)
df = next(df_iter)
len(df)

30000

#### Exploring our data.

In [37]:
df.head()

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0         2  2019-09-01 00:10:53   2019-09-01 00:23:46                  N   
1         2  2019-09-01 00:31:22   2019-09-01 00:44:37                  N   
2         2  2019-09-01 00:50:24   2019-09-01 01:03:20                  N   
3         2  2019-09-01 00:27:06   2019-09-01 00:33:22                  N   
4         2  2019-09-01 00:43:23   2019-09-01 00:59:54                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0           1            65           189                5           2.00   
1           1            97           225                5           3.20   
2           1            37            61                5           2.99   
3           1           145           112                1           1.73   
4           1           112           198                1           3.42   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0         10.5    0.5      0.5        2.36           0.0        NaN   
1         12.0    0.5      0.5        0.00           0.0        NaN   
2         12.0    0.5      0.5        0.00           0.0        NaN   
3          7.5    0.5      0.5        1.50           0.0        NaN   
4         14.0    0.5      0.5        3.06           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3         14.16             1          1   
1                    0.3         13.30             2          1   
2                    0.3         13.30             2          1   
3                    0.3         10.30             1          1   
4                    0.3         18.36             1          1   

   congestion_surcharge  
0                   0.0  
1                   0.0  
2                   0.0  
3                   0.0  
4                   0.0

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   VendorID               30000 non-null  int64  
 1   lpep_pickup_datetime   30000 non-null  object 
 2   lpep_dropoff_datetime  30000 non-null  object 
 3   store_and_fwd_flag     30000 non-null  object 
 4   RatecodeID             30000 non-null  int64  
 5   PULocationID           30000 non-null  int64  
 6   DOLocationID           30000 non-null  int64  
 7   passenger_count        30000 non-null  int64  
 8   trip_distance          30000 non-null  float64
 9   fare_amount            30000 non-null  float64
 10  extra                  30000 non-null  float64
 11  mta_tax                30000 non-null  float64
 12  tip_amount             30000 non-null  float64
 13  tolls_amount           30000 non-null  float64
 14  ehail_fee              0 non-null      float64
 15  im

#### We can see that our dataset column `lpep_pickup_datetime` & `lpep_dropoff_datetime` has the data into datetime format but the dataset has recognized it as a object, so we need to change them into `datetime` format.

In [39]:
df.lpep_pickup_datetime = pd.to_datetime(df.lpep_pickup_datetime)
df.lpep_dropoff_datetime = pd.to_datetime(df.lpep_dropoff_datetime)

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               30000 non-null  int64         
 1   lpep_pickup_datetime   30000 non-null  datetime64[ns]
 2   lpep_dropoff_datetime  30000 non-null  datetime64[ns]
 3   store_and_fwd_flag     30000 non-null  object        
 4   RatecodeID             30000 non-null  int64         
 5   PULocationID           30000 non-null  int64         
 6   DOLocationID           30000 non-null  int64         
 7   passenger_count        30000 non-null  int64         
 8   trip_distance          30000 non-null  float64       
 9   fare_amount            30000 non-null  float64       
 10  extra                  30000 non-null  float64       
 11  mta_tax                30000 non-null  float64       
 12  tip_amount             30000 non-null  float64       
 13  t

#### Now we can see our `dtype` for the column `lpep_pickup_datetime` & `lpep_dropoff_datetime` has been changed into `datetime`.

#### Creating the schema for our table through engine.

In [41]:
print(pd.io.sql.get_schema(df, name='green_taxi_data', con=engine))


CREATE TABLE green_taxi_data (
	"VendorID" BIGINT, 
	lpep_pickup_datetime TIMESTAMP WITHOUT TIME ZONE, 
	lpep_dropoff_datetime TIMESTAMP WITHOUT TIME ZONE, 
	store_and_fwd_flag TEXT, 
	"RatecodeID" BIGINT, 
	"PULocationID" BIGINT, 
	"DOLocationID" BIGINT, 
	passenger_count BIGINT, 
	trip_distance FLOAT(53), 
	fare_amount FLOAT(53), 
	extra FLOAT(53), 
	mta_tax FLOAT(53), 
	tip_amount FLOAT(53), 
	tolls_amount FLOAT(53), 
	ehail_fee FLOAT(53), 
	improvement_surcharge FLOAT(53), 
	total_amount FLOAT(53), 
	payment_type BIGINT, 
	trip_type BIGINT, 
	congestion_surcharge FLOAT(53)
)




#### Writing the first 10000 datas to the table.

#### Since this is the first batch of our data we are using `if_exists=replace` command to overwrite any data if exists before.

In [42]:
%time df.to_sql(name="green_taxi_data", con=engine, if_exists="replace", index=False)

CPU times: user 3.04 s, sys: 0 ns, total: 3.04 s
Wall time: 4.95 s


1000

#### Writing the remaining datas to the table.

In [43]:
from time import time

while True:
    try:
        t_start = time()
    
        df = next(df_iter)
        df.lpep_pickup_datetime = pd.to_datetime(df.lpep_pickup_datetime)
        df.lpep_dropoff_datetime = pd.to_datetime(df.lpep_dropoff_datetime)

        # Since the first batch has already been inerted we are using append option for further batches.
        df.to_sql(name='green_taxi_data', con=engine, if_exists='append', index=False)
    
        t_end = time()
    
        print("Inserted another chunk, took %.3f second" % (t_end - t_start))
    except Exception as ex:
            print(ex)
            break

Inserted another chunk, took 5.329 second
Inserted another chunk, took 5.011 second
Inserted another chunk, took 5.092 second
Inserted another chunk, took 5.259 second
Inserted another chunk, took 5.034 second
Inserted another chunk, took 5.410 second
Inserted another chunk, took 5.255 second
Inserted another chunk, took 5.242 second
Inserted another chunk, took 5.532 second
Inserted another chunk, took 5.132 second
Inserted another chunk, took 5.163 second
Inserted another chunk, took 5.513 second
Inserted another chunk, took 4.659 second
Inserted another chunk, took 4.633 second



### Code for writing data on single attempt:
```python
df = pd.read_csv("green_tripdata_2019-09.csv.gz", low_memory=False)
df.lpep_pickup_datetime = pd.to_datetime(df.lpep_pickup_datetime)
df.lpep_dropoff_datetime = pd.to_datetime(df.lpep_dropoff_datetime)
print(pd.io.sql.get_schema(df, name='green_taxi_data', con=engine))
df.to_sql(name="green_taxi_data", con=engine, if_exists="replace", index=False)
```
**Since we had already inserted data we're not gonna run this code: This is for ref purpose only.**

### Accessing data from the jupyter notebook.
For this purpose we are gonna use the [`ipython-sql`](https://github.com/catherinedevlin/ipython-sql) module.
For installation of the module we use the following command:
``` python
pip install ipython-sql
pip install psycopg2-binary
```
##### Load the extension with the command `%load_ext`

In [1]:
%load_ext sql

# Credentials for connecting to db
user_name = "root"
password = "root"
database_name = "ny_taxi"
host = "localhost"
port = "5432"

# Connection string format
connection_string = f"postgresql://{user_name}:{password}@{host}:{port}/{database_name}"

# Use the connection string with %sql magic
%sql $connection_string

In [50]:
%sql SELECT COUNT(1) FROM green_taxi_data;

 * postgresql://root:***@localhost:5432/ny_taxi
1 rows affected.


[(449063,)]

In [51]:
%sql SELECT * FROM green_taxi_data LIMIT 5;

 * postgresql://root:***@localhost:5432/ny_taxi
5 rows affected.


[(2, datetime.datetime(2019, 9, 1, 0, 10, 53), datetime.datetime(2019, 9, 1, 0, 23, 46), 'N', 1, 65, 189, 5, 2.0, 10.5, 0.5, 0.5, 2.36, 0.0, None, 0.3, 14.16, 1, 1, 0.0),
 (2, datetime.datetime(2019, 9, 1, 0, 31, 22), datetime.datetime(2019, 9, 1, 0, 44, 37), 'N', 1, 97, 225, 5, 3.2, 12.0, 0.5, 0.5, 0.0, 0.0, None, 0.3, 13.3, 2, 1, 0.0),
 (2, datetime.datetime(2019, 9, 1, 0, 50, 24), datetime.datetime(2019, 9, 1, 1, 3, 20), 'N', 1, 37, 61, 5, 2.99, 12.0, 0.5, 0.5, 0.0, 0.0, None, 0.3, 13.3, 2, 1, 0.0),
 (2, datetime.datetime(2019, 9, 1, 0, 27, 6), datetime.datetime(2019, 9, 1, 0, 33, 22), 'N', 1, 145, 112, 1, 1.73, 7.5, 0.5, 0.5, 1.5, 0.0, None, 0.3, 10.3, 1, 1, 0.0),
 (2, datetime.datetime(2019, 9, 1, 0, 43, 23), datetime.datetime(2019, 9, 1, 0, 59, 54), 'N', 1, 112, 198, 1, 3.42, 14.0, 0.5, 0.5, 3.06, 0.0, None, 0.3, 18.36, 1, 1, 0.0)]

In [74]:
%%sql
SELECT 
    CAST(lpep_dropoff_datetime AS DATE) AS "day",
    COUNT(1)
FROM
    green_taxi_data
GROUP BY
    CAST(lpep_dropoff_datetime AS DATE)
ORDER BY day ASC
LIMIT 100;

 * postgresql://root:***@localhost:5432/ny_taxi
51 rows affected.


[(datetime.date(2009, 1, 1), 10),
 (datetime.date(2009, 1, 5), 1),
 (datetime.date(2010, 9, 23), 4),
 (datetime.date(2019, 8, 31), 9),
 (datetime.date(2019, 9, 1), 12522),
 (datetime.date(2019, 9, 2), 10016),
 (datetime.date(2019, 9, 3), 14375),
 (datetime.date(2019, 9, 4), 15466),
 (datetime.date(2019, 9, 5), 15965),
 (datetime.date(2019, 9, 6), 16573),
 (datetime.date(2019, 9, 7), 15289),
 (datetime.date(2019, 9, 8), 13268),
 (datetime.date(2019, 9, 9), 14234),
 (datetime.date(2019, 9, 10), 14673),
 (datetime.date(2019, 9, 11), 15544),
 (datetime.date(2019, 9, 12), 16941),
 (datetime.date(2019, 9, 13), 16619),
 (datetime.date(2019, 9, 14), 15599),
 (datetime.date(2019, 9, 15), 13174),
 (datetime.date(2019, 9, 16), 14669),
 (datetime.date(2019, 9, 17), 14751),
 (datetime.date(2019, 9, 18), 15751),
 (datetime.date(2019, 9, 19), 16375),
 (datetime.date(2019, 9, 20), 16966),
 (datetime.date(2019, 9, 21), 16405),
 (datetime.date(2019, 9, 22), 13650),
 (datetime.date(2019, 9, 23), 14888),
 (datetime.date(2019, 9, 24), 15339),
 (datetime.date(2019, 9, 25), 16096),
 (datetime.date(2019, 9, 26), 16500),
 (datetime.date(2019, 9, 27), 16693),
 (datetime.date(2019, 9, 28), 15106),
 (datetime.date(2019, 9, 29), 12638),
 (datetime.date(2019, 9, 30), 12822),
 (datetime.date(2019, 10, 1), 108),
 (datetime.date(2019, 10, 10), 2),
 (datetime.date(2019, 10, 11), 1),
 (datetime.date(2019, 10, 12), 1),
 (datetime.date(2019, 10, 31), 2),
 (datetime.date(2019, 11, 2), 3),
 (datetime.date(2019, 11, 3), 3),
 (datetime.date(2019, 11, 21), 1),
 (datetime.date(2019, 11, 23), 2),
 (datetime.date(2019, 11, 24), 1),
 (datetime.date(2019, 11, 25), 1),
 (datetime.date(2019, 12, 1), 1),
 (datetime.date(2019, 12, 13), 1),
 (datetime.date(2019, 12, 15), 2),
 (datetime.date(2020, 1, 3), 1),
 (datetime.date(2020, 2, 15), 1),
 (datetime.date(2020, 3, 7), 1)]

* `DATE_TRUNC` is a function that trunctates a timestamp. When using `DAY` as a parameter, it removes any smaller values (hours, minutes, seconds) and displays them as `00:00:00` instead.
* `CAST` function will display the datas as simple dates removing the item `00:00:00`

## Question 3. Count records
**How many taxi trips were totally made on September 18th 2019?**

Tip: started and finished on 2019-09-18.

In [78]:
%%sql
SELECT
  COUNT(1)
FROM
  green_taxi_data
WHERE
  (lpep_pickup_datetime>='2019-09-17 00:00:00' AND
  lpep_pickup_datetime<'2019-09-18 00:00:00');

 * postgresql://root:***@localhost:5432/ny_taxi
1 rows affected.


[(14789,)]

In [79]:
%%sql
WITH daily_counts AS (
  SELECT
    CAST(lpep_dropoff_datetime AS DATE) AS "day",
    COUNT(1) AS "count"
  FROM
    green_taxi_data
  GROUP BY
    CAST(lpep_dropoff_datetime AS DATE)
)
SELECT count
FROM daily_counts
WHERE day = '2019-09-18';

 * postgresql://root:***@localhost:5432/ny_taxi
1 rows affected.


[(15751,)]

In [81]:
%%sql
SELECT 
    CAST(lpep_dropoff_datetime AS DATE) AS "day",
    COUNT(1)
FROM
    green_taxi_data
WHERE
    CAST(lpep_dropoff_datetime AS DATE) = '2019-09-18'
GROUP BY
    CAST(lpep_dropoff_datetime AS DATE);

 * postgresql://root:***@localhost:5432/ny_taxi
1 rows affected.


[(datetime.date(2019, 9, 18), 15751)]

In [72]:
%%sql
SELECT COUNT(*) AS trip_count
FROM green_taxi_data gt
WHERE CAST(gt.lpep_pickup_datetime AS DATE) = '2019-09-18'
  AND CAST(gt.lpep_dropoff_datetime AS DATE) = '2019-09-18';

 * postgresql://root:***@localhost:5432/ny_taxi
1 rows affected.


[(15612,)]

## Question 4. Largest trip for each day
Which was the pick up day with the largest trip distance Use the pick up time for your calculations.

In [2]:
%%sql
SELECT 
    CAST(lpep_pickup_datetime AS DATE), 
    MAX(trip_distance)
FROM
    green_taxi_data
GROUP BY CAST(lpep_pickup_datetime AS DATE)
ORDER BY 2 DESC LIMIT 1;

 * postgresql://root:***@localhost:5432/ny_taxi
1 rows affected.


[(datetime.date(2019, 9, 26), 341.64)]

## Question 5. Three biggest pick up Boroughs
Consider lpep_pickup_datetime in '2019-09-18' and ignoring Borough has Unknown

Which were the 3 pick up Boroughs that had a sum of total_amount superior to 50000?

In [99]:
%%sql
SELECT
    tz."Borough",
    SUM(total_amount) AS total_amount
FROM
    green_taxi_data gt
JOIN taxi_zones tz
ON gt."PULocationID" = tz."LocationID"
GROUP BY tz."Borough"
HAVING SUM(total_amount) > 50000
ORDER BY total_amount DESC
LIMIT 3;

 * postgresql://root:***@localhost:5432/ny_taxi
3 rows affected.


[('Brooklyn', 2619378.53999982),
 ('Queens', 2460386.170000191),
 ('Manhattan', 2427880.920000595)]

In [73]:
%%sql
SELECT
    tz."Borough" AS pickup_borough,
    SUM(gt.total_amount) AS total_amount
FROM
    green_taxi_data gt
JOIN
    taxi_zones tz ON gt."PULocationID" = tz."LocationID"
WHERE
    DATE(gt.lpep_pickup_datetime) = '2019-09-18'
    AND tz."Borough" != 'Unknown'
GROUP BY
    tz."Borough"
HAVING
    SUM(gt.total_amount) > 50000
ORDER BY
    total_amount DESC
LIMIT 3;


 * postgresql://root:***@localhost:5432/ny_taxi
3 rows affected.


[('Brooklyn', 96333.24000000033),
 ('Manhattan', 92271.2999999995),
 ('Queens', 78671.70999999894)]

## Question 6. Largest tip
For the passengers picked up in September 2019 in the zone name Astoria which was the drop off zone that had the largest tip? We want the name of the zone, not the id.

In [69]:
%%sql
SELECT 
    tz_dropoff."Zone" AS dropoff_zone_name,
    MAX(gt.tip_amount) AS largest_tip_amount
FROM green_taxi_data gt
JOIN taxi_zones tz_pickup ON gt."PULocationID" = tz_pickup."LocationID"
JOIN taxi_zones tz_dropoff ON gt."DOLocationID" = tz_dropoff."LocationID"
WHERE TO_CHAR(gt.lpep_pickup_datetime, 'YYYY-MM') = '2019-09'
    AND tz_pickup."Zone" = 'Astoria'
GROUP BY tz_dropoff."Zone"
ORDER BY largest_tip_amount DESC
LIMIT 1;

 * postgresql://root:***@localhost:5432/ny_taxi
1 rows affected.


[('JFK Airport', 62.31)]